### DISCUSSION TITLE
`Executor` Improvements

### DISCUSSION THREAD

**Discussion by LayneHaber** [connext/nxtp#752] [Archive]
## Summary
Remove the `ReentrancyGuard` for gas savings, no longer necessary with the `onlyTransactionManager` modifier. Update the order of actions. Adopt the contract calculation of `relayerFee` outlined [here](https://github.com/connext/nxtp/discussions/748).

## Motivation
There are a few improvements to the `FulfillInterpreter` that were recommended during the last audit:

1. [Order of actions in failure case](https://consensys.net/diligence/audits/2021/07/connext-nxtp-noncustodial-xchain-transfer-protocol/#fulfillinterpreter-wrong-order-of-actions-in-fallback-handling).
2. [Reentrancy Guard](https://consensys.net/diligence/audits/2021/07/connext-nxtp-noncustodial-xchain-transfer-protocol/#fulfillinterpreter-reentrancyguard-can-be-removed).

Both are minor issues, and have gotten different feedback from different auditors, but should be addressed regardless.

Additionally, calculating the appropriate `relayerFees` necessary for executing arbitrary calldata is complex. The routers only care if they are relaying the transaction, but relayer networks will start to enforce these are sensible as well.

## Proposed Solution
**Failure Case**
- Decrease the allowance before transferring the funds in the case where external calls fail

**Reentrancy**
- Remove the `ReentrancyGuard` and rely solely on the `onlyTransactionManager` modifier

**Gas Costs**
- The user needs to know when they `prepare` the transaction what the approximate cost of the execution of `calldata` on the receiving chain will be. This is because they must include enough in the `SenderPrepare` to cover the router fees, relayer fees, and be sufficient to get across chains
- Because the `calldata` is encrypted, nobody else can assess this ahead of time. Additionally, it may fail on `.estimateGas` calls until the `Fulfill` is done.
- Ultimately, this is within the purview of the user/contract executing the `SenderPrepare`. We should allow the user to pass in the `gasPrice` and the `gasLimit` for a given `fulfill` transaction to the sdk functions, and continue to emit the `relayerFee` used in events.

## Test Cases

Not sure what test cases could be included to assess these changes outside of the expected scope of other tests.

## Tasks

**Comment by liu-zhipeng** [comment]
I think that is a good idea to remove `ReentrancyGuard `. which was not necessary with the `onlyTransactionManager ` at all.

**Reply by LayneHaber** [reply]
Agreed! It was an oversight.

**Comment by LayneHaber** [comment]
Updated to reflect that #748 will not calculate costs onchain.

**Comment by LayneHaber** [comment]
Implemented in #865
